src/dune_rules/melange_rules.ml

```diff
@@ -0,0 +1,79 @@
+open Import
+
+let gen_rules ~melange_stanza_dir ~scope ~sctx ~expander
+    (mel : Melange_stanzas.Emit.t) =
+  let all_libs_compile_info =
+    let dune_version = Scope.project scope |> Dune_project.dune_version in
+    let pps = [] in
+    Lib.DB.resolve_user_written_deps_for_exes (Scope.libs scope)
+      [ (mel.loc, mel.target) ]
+      mel.libraries ~pps ~dune_version
+  in
+  let lib_cctx ~sctx ~obj_dir ~modules ~expander ~scope ~compile_info =
+    let flags = Ocaml_flags.empty in
+    let vimpl =
+      (* original impl in Lib_rules uses Virtual_rules.impl, will this break with virtual libs? *)
+      None
+    in
+    let modules = Vimpl.impl_modules vimpl modules in
+    let requires_compile = Lib.Compile.direct_requires compile_info in
+    let requires_link = Lib.Compile.requires_link compile_info in
+    let package = None in
+    let js_of_ocaml = None in
+    let melange =
+      Melange.In_context.make ~target:mel.target ~melange_stanza_dir
+    in
+    (* modes and pp are not passed, not sure if this will cause issues *)
+    Compilation_context.create () ~super_context:sctx ~expander ~scope ~obj_dir
+      ~modules ~flags ~requires_compile ~requires_link
```

Comment:
Thinking about this a little more, I don't think you even need a Cctx. The kind of "compilation" you're doing is too different from regular OCaml compilation so very little is gained by reusing Cctx.

Reply:
Let's defer this entry_module stuff for later. It's totally optional at this point.

Reply:
I removed most usages of `cctx` in https://github.com/ocaml/dune/pull/6349/commits/648892205b7af6f1c9d4a1ffa4dedee61f03aed4. 

There is one case left, where I am reusing `Module_compilation.build_all` to get the cmj rules for the entries. I will see if there's a way around it.

Reply:
Actually, with recent approach, using `cctx` makes more sense as the modules in `melange.emit` need to get their cmj rules generated, so having cctx allows to call `build_all` (same as with exes):

https://github.com/jchavarri/dune/blob/f0bdba3c0a0fdd7affc6d5daf85c142083230b42/src/dune_rules/melange_rules.ml#L134

Reply:
Yes, you need the compilation context to compile your entry modules to cmj, but you shouldn't use it when compiling cmj to js